Sources/DistributedSystem/ChannelCompression.swift

```diff
@@ -168,102 +490,208 @@ fileprivate struct BufferManager: ~Copyable {
     }
 
     func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+        fatalError("should never be called")
+    }
+}
+
+final class ChannelStreamCompressionInboundHandler: ChannelCompressionInboundHandler {
+    private var lz4Stream = LZ4_streamDecode_t()
+    private var bufferManager = BufferManager()
+
+    override init(_ distributedSystem: DistributedSystem) {
+        LZ4_setStreamDecode(&lz4Stream, nil, 0)
+        super.init(distributedSystem)
+    }
+
+    override func channelRead(context: ChannelHandlerContext, data: NIOAny) {
         var bufferIn = unwrapInboundIn(data)
         _ = bufferIn.readInteger(as: UInt32.self) // skip message size
-        if let messageSize = bufferIn.readInteger(as: UInt32.self) {
-            let dstBuffer = bufferManager.getNextBuffer(capacity: Int(messageSize))
-            let decompressedBytes = bufferIn.withUnsafeReadableBytes { ptr in
+        do {
+            let (messageSizeSize, messageSize) = try bufferIn.withUnsafeReadableBytes { try ULEB128.decode($0, as: UInt32.self) }
+            bufferIn.moveReaderIndex(forwardBy: messageSizeSize)
+            let buffer = bufferManager.getNextBuffer(capacity: Int(messageSize))
+            let decompressedBytes = bufferIn.withUnsafeReadableBytes { bytes in
                 LZ4_decompress_safe_continue(&lz4Stream,
-                                             ptr.baseAddress!.assumingMemoryBound(to: Int8.self),
-                                             dstBuffer.baseAddress!.assumingMemoryBound(to: Int8.self),
-                                             Int32(ptr.count),
-                                             Int32(dstBuffer.count))
+                                             bytes.baseAddress!.assumingMemoryBound(to: Int8.self),
+                                             buffer.baseAddress!.assumingMemoryBound(to: Int8.self),
+                                             Int32(bytes.count),
+                                             Int32(buffer.count))
             }
             if decompressedBytes == messageSize {
                 var bufferOut = ByteBufferAllocator().buffer(capacity: MemoryLayout<UInt32>.size + Int(messageSize))
                 bufferOut.writeInteger(messageSize)
                 _ = bufferOut.writeWithUnsafeMutableBytes(minimumWritableBytes: 0) { ptr in
-                    ptr.copyMemory(from: UnsafeRawBufferPointer(start: dstBuffer.baseAddress, count: Int(messageSize)))
+                    ptr.copyMemory(from: UnsafeRawBufferPointer(start: buffer.baseAddress, count: Int(messageSize)))
                     return Int(messageSize)
                 }
                 context.fireChannelRead(wrapInboundOut(bufferOut))
                 bytesDecompressed += UInt64(messageSize)
                 return
             }
+        } catch {
         }
         logger.info("\(context.channel.addressDescription): invalid compressed message received, close connection")
```

Comment:
It is really unexpected the message will not be parsed properly here because it passed previous handler in pipeline where data stream is defragmented to messages. We could spend some time to extend error logging but I am not sure it will ever will be triggered.